pkg/conversion/cli.go

```diff
@@ -0,0 +1,272 @@
+package conversion
+
+import (
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/json"
+
+	xpmeta "github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+
+	"github.com/crossplane-contrib/terrajet/pkg/meta"
+	"github.com/crossplane-contrib/terrajet/pkg/terraform/resource"
+	"github.com/crossplane-contrib/terrajet/pkg/tfcli"
+)
+
+const (
+	errCannotGetClientBuilder = "cannot get client builder"
+	errCannotConsumeState     = "cannot consume state"
+
+	errFmtFailedToWithTFCli = "failed to %s with tf cli"
+	errFmtCannotBuildClient = "cannot build %s client"
+)
+
+// Cli is an Adapter implementation for Terraform Cli
+type Cli struct {
+	builderBase tfcli.Builder
+}
+
+// NewCli returns a Cli object
+func NewCli(cliBuilder tfcli.Builder) *Cli {
+	return &Cli{
+		builderBase: cliBuilder,
+	}
+}
+
+// Observe is a Terraform Cli implementation for Observe function of Adapter interface.
+func (t *Cli) Observe(tr resource.Terraformed) (ObserveResult, error) {
+	b, err := t.getClientBuilderForResource(tr)
+	if err != nil {
+		return ObserveResult{}, errors.Wrap(err, errCannotGetClientBuilder)
+	}
+	tfc, err := b.BuildObserveClient()
+	if err != nil {
+		return ObserveResult{}, errors.Wrapf(err, errFmtCannotBuildClient, "observe")
+	}
+
+	// Attempt to make an observation. There are a couple of possibilities at this point:
+	// a. No tfcli operation in progress, we just kick off a new observation. It should
+	//    immediately return "tfRes.Completed" as "false", and we return completed=false in ObserveResult.
+	// b. An "observe" operation is in progress that we kicked off in one of the previous reconciliations.
+	//    This call would return tfRes.Completed as false, and we would return completed=false in ObserveResult.
+	// c. A previously started "observe" operation completed. We can just consume state and return ObserveResult
+	//    accordingly.
+	// d. A previously started "create" operation is in progress or completed but its state needs to be
+	//    read to kick off a new operation. We will return "Exists: false" in order to trigger a Create call.
+	// e. A previously started "update" operation is in progress or completed but its state needs to be
+	//    read to kick off a new operation. We will return "UpToDate: false" in order to trigger an Update call.
+	// f. A previously started "delete" operation is in progress and by returning "Exists: true" and since
+	//    deletion timestamp should already be set, it would trigger a delete call.
+	tfRes, err := tfc.Observe(xpmeta.GetExternalName(tr))
```

Comment:
I see your point and the possible benefits of making observe call blocking. 
However, as I stated in my comment [here](https://github.com/crossplane-contrib/terrajet/pull/14#discussion_r691579728), this would mean being able to run parallel commands (`Show` + `Apply` or `Show` + `Destroy`) on a single resource. If there are no blockers on this regarding tfcli implementation, I would agree that this would simplify the logic here.

Reply:
We discussed this offline, just writing it down here. What I'm proposing is that when `ExternalClient` asks for an observation we either (1) return immediately saying Apply/Delete is in progress (which is already handled by reporting existence/uptodate etc.) or (2) call `terraform refresh/show` and block until it returns. So, at a high level we make an observation and report that back. That's why I was pushing for not considering being `InProgress` as an error state because it's similar to other controllers where `Describe` call can return that the resource is still being created and that's not an error. One can think of our `Cli` tool as the cloud provider client here; we're not really concerned what's blocking or not we just want observation of the world.

Reply:
I agree that we can provide a synchronous `BlockingObserve` call in `tfcli.Client`. Initially we can also keep the async interface.

Reply:
I just checked the implementation in provider-azure. A blocking observe call is also the case there and the azure-sdk-for-go library does not return a future for the `Get` call.

Reply:
Updated accordingly, as we discussed offline 👍